### PR TITLE
Speed up MetadataStateFormat Writes

### PR DESCRIPTION
**Pull request by original-brownbear**
There's some needless deleting and exists checking here that gets quite expensive
in tests and in some benchmarks. We can save quite a few system calls
and IO by doing what `FsBlobContainer` does and optimistically attempt
writes that will essentially always work out.
The deletes of the temp paths are also noops unless there's an exception so we can
skip those as well.
Lastly, `Path.resolve` isn't free or even cheap in some cases, so I added some guards
around trace logging for it.

This speeds up tests by a couple of seconds on each `:server:check` run and saves IOPS
in production as well.

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/gateway/MetadataStateFormat.java [hunk 1]
```diff
@@ -95,25 +98,12 @@ private static void deleteFileIgnoreExceptions(Path stateLocation, Directory dir
...
-                    }
-                })) {
-                    builder.startObject();
-                    toXContent(builder, state);
-                    builder.endObject();
-                }
-                CodecUtil.writeFooter(out);
+            try {
+                doWriteToFirstLocation(state, stateDir, tmpFileName);
+            } catch (FileAlreadyExistsException fae) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/gateway/MetadataStateFormat.java:103]: It doesn't look like anything in `MetadataStateFormatTests` exercises this `catch`. Could you add a test that does?

File: server/src/test/java/org/elasticsearch/gateway/MetadataStateFormatTests.java [hunk 2]
```diff
@@ -280,7 +280,6 @@ public void testFailWriteAndReadPreviousState() throws IOException {
 
         for (int i = 0; i < randomIntBetween(1, 5); i++) {
             format.failOnMethods(
-                Format.FAIL_DELETE_TMP_FILE,
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/gateway/MetadataStateFormatTests.java:283]: Would be nice to still check that if we need to delete a file and the delete fails then we do the right thing.

**Review by DaveCTurner — COMMENTED** [review]
Looks ok to me, I asked for a few test changes tho.

File: server/src/test/java/org/elasticsearch/gateway/MetadataStateFormatTests.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/test/java/org/elasticsearch/gateway/MetadataStateFormatTests.java:283]: Sure thing added some infrastructure for this and a test that simulates the delete failure and the catch above in one (with the auto-cleanup on this path I needed both a write and delete failure in one anyway :)).

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**